feat(wandb): log models as artifacts

borisdayma opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
 Please also include relevant motivation and context.
 List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->
Fixes #4903

should be merged after #5931 

## Before submitting
- [X] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [X] Did you make sure to update the documentation with your changes? (if necessary)
- [X] Did you write any new necessary tests? (not for typos and docs)
- [X] Did you verify new and existing tests pass locally with your changes?
- [X] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- For CHANGELOG separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @borisdayma! Thanks for opening this PR. 

* In the file [`pytorch_lightning/loggers/wandb.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/3fb55f66b57f4f3283d98ee1282a7304d1984097/pytorch_lightning/loggers/wandb.py):

> [Line 220:106](https://github.com/PyTorchLightning/pytorch-lightning/blob/3fb55f66b57f4f3283d98ee1282a7304d1984097/pytorch_lightning/loggers/wandb.py#L220): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace







Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

borisdayma commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -209,6 +207,18 @@ def version(self) -> Optional[str]:
+            }
+            # TODO: see if we can also log data from `trainer.checkpoint_callback` (best_model_path, etc) 
+            artifact = wandb.Artifact(name=f"run-{artifact_name}", type="model", metadata=metadata)
+            # TODO: we need access to `trainer.checkpoint_callback.dirpath`
```
@awaelchli  Do you know how I could access `train.checkpoint_callback.dirpath`?
I just need to know where the models are located. If I can access the entire callback, I could actually also log interesting values (best_model_path, etc).

awaelchli replied:
hmm no, the trainer is unfortunately nowhere accessible in the logger. A possible solution is to pass the trainer object to the finalize method. This would have to be updated in a few other places and loggers though.

borisdayma replied:
What about passing it at init so loggers can access all items they need at that point (or save a reference)?

awaelchli replied:
Not in love with this idea tbh. I believe our wrappers around the 3rd party loggers should also be usable without a trainer. 
How about a connect method: At the appropriate time, say in fit, trainer calls
`self.logger.connect(self)` passes itself or maybe also the model, like `self.logger.connect(self, model)`. 
The logger can then read required attributes from trainer.

borisdayma replied:
Yes that would work. I imagine it will iterate over all the loggers (or also the callbacks)?

What about just passing a reference to ModelCheckpoint (if the objective is to keep the Trainer separate from loggers) which defaults to None?

justusschock replied:
@borisdayma I think a connect method works best (would probably be called by `LoggerConnector`).

If you give it a go, it should be fairly easy to implement the baseclass method for it to do actually nothing (just for signature consistency).

I would recommend to only pass in the trainer and preferably only save a `weakref` or proxy to it, since loggers should not have a reference that keeps it alive :)

Depending on future other tasks the `ModelCheckpoint may not be sufficient and for other parts we also pass the trainer there (just not upon init so it's optional).

borisdayma replied:
So I'm looking at a few places based on your suggestions where this connection can be made. We could update `LoggerConnector.configure_logger` (which attaches logger to the trainer):
1. we just add the reverse operation, attaching the trainer to the logger (as a weakref or proxy)
2. we just add a call to `logger.connect` function which does nothing by default

Does one of these methods sound reasonable?

justusschock replied:
I would go with option 2

borisdayma commented:
I started from my other ongoing PR so you can see the changes [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/6231/commits/3fb55f66b57f4f3283d98ee1282a7304d1984097)

borisdayma commented:
The history looks a bit messed up so I'll try to fix it or force push from a clean master state (since there's only one real commit so far).

justusschock commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -215,6 +213,18 @@ def version(self) -> Optional[str]:
+            # TODO: see if we can also log data from `trainer.checkpoint_callback` (best_model_path, etc) 
+            artifact = wandb.Artifact(name=f"run-{artifact_name}", type="model", metadata=metadata)
+            # TODO: we need access to `trainer.checkpoint_callback.dirpath`
+            artifact.add_dir(trainer.checkpoint_callback.dirpath)
```
does that mean you log all checkpoints inside that directory?

Wouldn't it be better to only log the relevant parts (accessible through `trainer.checkpoint_callback.kth_best_model_path`, `trainer.checkpoint_callback.best_model_path`, `trainer.checkpoint_callback.last_model_path` )? 

In the worst case you might be saving hundreds of checkpoints nobody needs.

borisdayma replied:
Great idea!

justusschock commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -158,6 +159,9 @@ def configure_logger(self, logger):
                 self.trainer.logger = logger
 
+        # connect trainer to logger
+        logger.connect(self.trainer)
```
can we maybe already make it a weakproxy here? We don't know if all the loggers will correctly have only a weakref on this otherwise.
```suggestion
        logger.connect(weakref.proxy(self.trainer))
```

justusschock commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -142,6 +149,10 @@ def __getstate__(self):
         state['_experiment'] = None
         return state
 
+    def connect(self, trainer: 'Trainer') -> None:
+        if trainer is not None:
+            self._trainer = proxy(trainer)
```
Should we maybe move the creation of the proxy already to the connector? That way we can ensure that no logger can actually bind to the real trainer object but only to the proxy.

borisdayma commented:
Here are examples:
* with default checkpointing: [artifact metadata](https://wandb.ai/borisd13/MNIST/artifacts/model/run-autumn-vortex-179/c1bd00cb9a9b46e55d4c/metadata)
* with custom `ModelCheckpoint`: [artifact metadata](https://wandb.ai/borisd13/MNIST/artifacts/model/run-sleek-dawn-183/878a2c0d4ead3930a942/metadata)

The idea is to avoid putting too much metadata that would hide more meaningful data.

borisdayma commented:
Here is current behavior:
* default: log only last model as artifact -> [example artifact](https://wandb.ai/borisd13/MNIST/artifacts/model/run-proud-planet-195/ac1273c8f656053fb838)
* custom `ModelCheckpoint`: according to options (looks through `last_model_path`, `best_model_path` and `best_k_models`) -> [example artifact](https://wandb.ai/borisd13/MNIST/artifacts/model/run-bright-frost-194/fecb31662233de3e16d5)

We can see in the 2nd example that 3 model were logged:
* versions are ordered by timestamp
* `latest` and `best` aliases are automatically set
* file within artifact is called `model.ckpt`

This would let users access the `best` model from a specific run with a command such as:
`artifact = run.use_artifact('USER/PROJECT/RUN_NAME:best', type='model')`

justusschock commented on pytorch_lightning/loggers/base.py:
```diff
@@ -71,6 +75,15 @@ def __init__(
         self._agg_key_funcs = agg_key_funcs if agg_key_funcs else {}
         self._agg_default_func = agg_default_func
 
+    def connect(self, trainer: 'Trainer') -> None:
```
Can you add the typing here to a proxy? I am not sure if this really works with typechecks against trainer otherwise. Or we use cast right after we created the proxy in the connector.

justusschock commented on pytorch_lightning/loggers/base.py:
```diff
@@ -355,6 +368,10 @@ def __init__(self, logger_iterable: Iterable[LightningLoggerBase]):
     def __getitem__(self, index: int) -> LightningLoggerBase:
         return [logger for logger in self._logger_iterable][index]
 
+    def connect(self, trainer: 'Trainer') -> None:
```
same with typing

justusschock commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -142,6 +145,10 @@ def __getstate__(self):
         state['_experiment'] = None
         return state
 
+    def connect(self, trainer: 'Trainer') -> None:
```
also typing :)

justusschock commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -142,6 +145,10 @@ def __getstate__(self):
         return state
 
+    def connect(self, trainer: 'Trainer') -> None:
+        if trainer is not None:
```
shouldn't this one always be not None if called at all?

justusschock commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -143,6 +144,7 @@ def should_update_logs(self):
         return should_log_every_n_steps or self.trainer.should_stop
 
     def configure_logger(self, logger):
+        # connect logger to trainer
```
```suggestion
```

justusschock reviewed (commented):
I like it. Just some more comments :) 

Will the artifacts be uploaded and updated continuously? E.g. will I at every point of the training have a best model artifact at wandb and if I get a new one it will replace the old one?

borisdayma commented:
> Will the artifacts be uploaded and updated continuously? E.g. will I at every point of the training have a best model artifact at wandb and if I get a new one it will replace the old one?

At the moment it would just happen at the end of training.

However it would be very useful to allow it to happen during training as well, maybe with something like:
* `WandbLogger(log_model = True)` to log at the end of training only whichever checkpoints we have (current behavior)
* `WandbLogger(log_model = 'all')` to log continuously

An easy way would be to add a call to `logger.after_save_checkpoint()` at the end of `ModelCheckpoint.save_checkpoint`.
What do you think?

borisdayma commented:
I experimented with using `logger.after_save_checkpoint` so that we could continuously log checkpoints.

* `WandbLogger(log_model=True)` will log checkpoints as the end of training according to `ModelCheckpoint` (so only the last one by default but it could change based on `save_top_k`)
* `WandbLogger(log_model='all')` will log any new checkpoint during training. Same if `log_model==True and save_top_k == -1`

Each model is uniquely saved as an artifact.
Aliases for `latest` and `best` are automatically set so you could easily recover those.

Here is [a demo notebook](https://colab.research.google.com/drive/17SaJQOTJuz5vV4tD0YEpk_Oy7dvKhvIO?usp=sharing).

justusschock commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -213,8 +219,44 @@ def version(self) -> Optional[str]:
         # don't create an experiment if we don't have one
         return self._experiment.id if self._experiment else self._id
 
+    @rank_zero_only
```
can we check this condition inside the trainer? That way it's easer to ensure that this was only called by rank 0. Otherwise this is an easy point of failure for new logger implementations

borisdayma replied:
Is it something that should be enforced?

I'm not sure of the logic but currently:
* the callback calls `ModelCheckpoint.save_checkpoint` which must run on all ranks and delegate proper responsibility to `self.save_function` (per docstring)
  * indirectly it calls `CheckpointConnector.save_checkpoint` which runs on rank 0
  * it also calls `logger.after_save_checkpoint` which does the same and run only on rank 0 (but maybe another logger would want something else?)

tchaton replied:
I agree with @justusschock. Better to call after_save_checkpoint only on rank 0 and put this logic in ModelCheckpoint. If a use case requires all ranks, we will update the code then.

borisdayma replied:
Sounds good. I made the change.

justusschock reviewed (approved):
I like it. Just one more question:

When you set `log_model='all`' and you get a new best checkpoint the old best one is still present right? it won't be deleted?
This is fine for 'all' but IMO with `log_model=True` we could use the same logic and just remove the old artifacts (not sure if this is possible with current wandb API).

borisdayma commented:
So the main idea between the 2 options was for people with limited bandwidth (due to speed or total monthly data).

If you have no issue with it you should just upload all models with `log_model='all'`.
Later in wandb you should be able to auto remove unused artifacts after a certain period, for example with no alias such as `best` or `latest` that are added here

justusschock commented:
@borisdayma I get it, but there may be an additional use case:

Limited data on wandb itself. That's why it would make sense to auto-remove the older best if there is a new one.

borisdayma commented:
That makes sense.
At the moment there doesn't seem to be an easy way to remove an artifact from the run API. I'll investigate further.
When it's possible we could easily add an option `log_model='best` for this suggested behavior.

I believe there's supposed to eventually be a setting to let you create a policy that would auto-remove unused artifacts.

tchaton commented on pytorch_lightning/loggers/wandb.py:
```diff
@@ -213,8 +233,42 @@ def version(self) -> Optional[str]:
+            checkpoint_callback.last_model_path: checkpoint_callback.current_score,
+            checkpoint_callback.best_model_path: checkpoint_callback.best_model_score,
+            **checkpoint_callback.best_k_models}
+        checkpoints = sorted([(Path(p).stat().st_mtime, p, s)
```
Hey, 2 small questions:
1. Do you see a need to use the time to filer previous checkpoint. Won't it be simpler to keep a list of upload checkpoint.
2. What happens if the training fails and the user re-starts from a given checkpoint.  `_logged_model_time` will be empty and all checkpoints will be re-uploaded.

borisdayma replied:
1. The issue is you could have checkpoints with the same name (eg: `last.ckpt`) so the time helps to know it's a new file.
2. If the training fails, I imagine you would start a new run (maybe you would even reload from a previous checkpoint), so you would want to upload the checkpoints. Each run is associated to a list of checkpoints contained within an artifact name. 

When you want to reload a model, you would choose the run and then the version of your artifact (typically the one we will call `best` with this integration). I'm currently writing [a demo](https://colab.research.google.com/drive/1Tr9dYlwBKk6-LgLKGO8KYZULnguVA992#scrollTo=4rFC9ICWazTy) so you can see what it does

tchaton commented on tests/loggers/test_wandb.py:
```diff
@@ -151,6 +143,43 @@ def test_wandb_logger_dirs_creation(wandb, tmpdir):
 
 
+@mock.patch('pytorch_lightning.loggers.wandb.wandb')
+def test_wandb_log_model(wandb, tmpdir):
```
Can we add a test for restarting an experiment ?

borisdayma replied:
I'm not sure I understand.
If you mean reloading a model, it is not done by the integration but by `wandb` api (see demo) and interact with the servers (to download the artifact). The logic of uploading a file and re-downloading the same file would probably be more specific to `wandb` itself.

tchaton reviewed (commented):
Great addition. Some questions about re-starting an experiment.

borisdayma commented:
I don't understand the failed tests but they seem unrelated to this PR.

Let me know if you need anything else from me!

borisdayma commented:
I updated default args logged per https://github.com/PyTorchLightning/pytorch-lightning/commit/cea170e011861d084ef16422d34a63c90fc72021 and added a new test.

I think it's now ready to be merged unless you have any other comments.

awaelchli commented on tests/loggers/test_wandb.py:
```diff
@@ -182,7 +229,7 @@ def wrapper_something():
 
 
-@mock.patch('pytorch_lightning.loggers.wandb.wandb')
+@ mock.patch('pytorch_lightning.loggers.wandb.wandb')
```
accident?

borisdayma replied:
Thanks for noticing!
I guess it didn't raise an error because no mock is needed here. I'll update.

borisdayma commented on pytorch_lightning/callbacks/model_checkpoint.py:
```diff
@@ -260,6 +261,10 @@ def save_checkpoint(self, trainer, unused: Optional = None):
         # Mode 3: save last checkpoints
         self._save_last_checkpoint(trainer, monitor_candidates)
 
+        # notify loggers
+        if trainer.is_global_zero and trainer.logger and hasattr(trainer.logger, 'after_save_checkpoint'):
+            trainer.logger.after_save_checkpoint(proxy(self))
```
The problem there is that the checkpoint has not been saved yet. We just gather states.
’after_save_checkpoint’ needs to happen after the model has been saved locally.

borisdayma replied:
@awaelchli Here is the stack when saving a checkpoint. Items at the same level are called sequentially.

* `ModelCheckpoint.save_checkpoint` → `_save_top_k_checkpoint` and `_save_none_monitor_checkpoint` and `_save_last_checkpoint` → `ModelCheckpoint._save_model` → `ModelCheckpoint._do_save` → `Trainer.save_checkpoint`
  * → `CheckpointConnector.save_checkpoint`
    * → `CheckpointConnect.dump_checkpoint` only gather states through `on_save_checkpoint`
    * → `atomic_save` saves the file

So basically if you prefer it in the checkpoint connector it could be at the end of `CheckpointConnector.save_checkpoint` (after `atomic_save`).

The issue is that it won't be as easy to get access to `ModelCheckpoint` from `logger.after_save_checkpoint` (which we need especially to access checkpoint path):
* calling `after_save_checkpoint` from `ModelCheckpoint` let us pass directly the `ModelCheckpoint` object
* calling `after_save_checkpoint` from `CheckpointConnector` requires a more complex logic, probably also connecting the trainer to the logger with `logger.connect` as we had previously

Let me know how you want me to proceed.

awaelchli replied:
I definitely think that it should not be the responsibility of the model checkpoint callback to manage this logging logic.
Instead, I propose to let the trainer/connector orchestrate this. How about this:

In ModelCheckpoint:

```python
def after_save_checkpoint(self, trainer):
     # here you can access trainer.checkpoint_callback

```

In CheckpointConnector:

```python
    def save_checkpoint(self, filepath, weights_only: bool = False):
        ...
        atomic_save(checkpoint, filepath)
        # after save:
        self.trainer.logger.after_save_checkpoint(self.trainer)
```

borisdayma replied:
I've been trying to move the logic to `CheckpointConnector.save_checkpoint` but ran into an issue with this approach.

The parameters from `ModelCheckpoint` ("best_model_path", "last_model_path", etc) are updated only later, at [these lines](https://github.com/PyTorchLightning/pytorch-lightning/blob/6a1414681107cf77d1ebeba86359df7b6668bf49/pytorch_lightning/callbacks/model_checkpoint.py#L280-L284).

Those are required to know if the checkpoint should be saved or not, and if the current score improved (to define which is the "best" checkpoint) so it needs to happen after (or at the end of) `ModelCheckpoint.save_checkpoint`.

awaelchli replied:
sorry, didn't think of that. I will help brainstorming and fixing the issue after iccv deadline on thursday. Sorry that this is blocking you

borisdayma replied:
No worry, I definitely want it to be implemented the best way so let me know if you want me to try another approach!

awaelchli replied:
I played around with this. I have an idea how to resolve it, but for this we would need to remember which checkpoints we have already uploaded and which not. That should be possible, right?

borisdayma replied:
Yes, I use `self._logged_model_time` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/fda377f6c63417ee8a8c6e4a519b8912d3e37044/pytorch_lightning/loggers/wandb.py#L157) for this purpose.

It captures the path as a key and timestamp as value (as we could have sometimes same file name with newer chekpoint).

Then the logic to find the new checkpoints to log is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/fda377f6c63417ee8a8c6e4a519b8912d3e37044/pytorch_lightning/loggers/wandb.py#L250-L258).

It may happen after we saved a checkpoint or only at the end of training (`log_model=True`) where we may have several checkpoints to log (for example with `save_top_k=3`).

awaelchli replied:
would the call to `wandb.log_artifact` upload the files asynchronoulsly?

borisdayma replied:
Yes!

borisdayma replied:
Just a note related to using `Callback`, we would still need a new hook as it needs to happen after saving the checkpoint (so `on_save_checkpoint` would not be the right place).

borisdayma replied:
Just checking for an update @awaelchli . I think many people are interested in this feature :smiley: 

Right now this current version is the cleanest way I found. Would it make sense to merge it and update later if there is some other functionality (logger considered as regular callbacks, new "after_save" method for callback, etc)?

awaelchli replied:
Sorry for letting you hang for a few days on this! Yes it is indeed a nice feature and I myself am a heavy wandb user. So certainly appreciate it.
I do not see a more elegant solution right now without larger changes, but I don't want to block it until we find one. 
@PyTorchLightning/core-contributors let us know what you think. In summary, the question is:
Is it acceptable that the `ModelCheckpoint` reaches through the Trainer into the logger to trigger a notification when saving is complete, so that the logger can upload the checkpoint files?

awaelchli replied:
If yes, I think we should go ahead with this PR. Uploading the model to the WandB cloud and keeping it versioned together with the logs is a very nice feature to have from the perspective of reproducibility

borisdayma replied:
Let me know if we should proceed and I can merge master to keep this branch in sync

awaelchli replied:
I will ping the team internally

borisdayma commented on docs/source/common/loggers.rst:
```diff
@@ -226,8 +231,8 @@ The :class:`~pytorch_lightning.loggers.WandbLogger` is available anywhere except
             some_img = fake_image()
-            self.logger.experiment.log({
-                 "generated_images": [wandb.Image(some_img, caption="...")]
+            self.log({
+                "generated_images": [wandb.Image(some_img, caption="...")]
             })
```
It actually does! I do it [here](https://colab.research.google.com/drive/1Tr9dYlwBKk6-LgLKGO8KYZULnguVA992?usp=sharing#scrollTo=2diE1gzlzpqZ).
Note : you may want to about the learning rate finder section due to current issue https://github.com/PyTorchLightning/pytorch-lightning/issues/6471

It presents some advantage when reloading the Trainer later as this logging becomes a noop until adding a logger (which would raise an issue calling ’self.logger.experiment...’).

The issue may be when several loggers are present, it will raise an error with the other ones.

Let me know if it's not supposed to be used this way and I'll update the doc and upcoming demo.

awaelchli replied:
ok, I didn't know that we support this

borisdayma commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -158,6 +159,10 @@ def configure_logger(self, logger):
+        # connect trainer to logger
+        if hasattr(self.trainer.logger, 'connect'):
+            self.trainer.logger.connect(self.trainer)
+
```
I think you commented on an older commit. We don't need to "connect" the trainer anymore.

borisdayma commented:
Pending items:
* discussion about where to call `after_save_checkpoint` with @awaelchli : https://github.com/PyTorchLightning/pytorch-lightning/pull/6231#discussion_r594805311
* update doc for logging

williamFalcon commented on docs/source/common/loggers.rst:
```diff
@@ -226,8 +231,8 @@ The :class:`~pytorch_lightning.loggers.WandbLogger` is available anywhere except
-            self.logger.experiment.log({
-                 "generated_images": [wandb.Image(some_img, caption="...")]
+            self.log({
+                "generated_images": [wandb.Image(some_img, caption="...")]
```
hey guys... we need to standardize for all loggers, not just wnb. let's sync up on this to make sure these changes aren't just for a single logger.

borisdayma replied:
Let me know if I need to update this on the other loggers as well.

I was just trying to take advantage of this PR to clean up the doc since I often get asked on when to use `self.log`, `self.logger.experiment.log` or even `self.logger[0].experiment.log` (I typically suggest to just try to use `self.log`).

tchaton replied:
`self.log_images` could be the API used to log every image related artefacts.

tchaton reviewed (approved):
LGTM !

borisdayma commented:
Hi guys, just checking if anything else is needed from me.
I know quite a few people already who are interested in accessing this feature.

edenlightning commented:
Apologize for the late reply Boris!
We want to spend some time thinking about how to make this a robust API that would work well for all supported loggers. We have 1.3 release coming soon which is too late to add this in, so we will prioritize this right after the release.

Thanks again for your contribution, and sorry we haven't been able to address quicker.

borisdayma commented:
Hi, I'm just checking if there's any update or if I can do anything from side.
It would be great to implement this feature as it will close a lot of current open issues (and make the integration so much better!)

borisdayma commented:
Hey guys, just wondering if you would be ok using my current version?

Right now I just need a callback for `after_save_checkpoint` so whenever you have a more robust one, it will be easy to switch it out.

Borda commented:
@awaelchli how is it going here, I saw you had most comments so mind giving your decision? :rabbit:
@borisdayma mind resolve conflicts?

borisdayma commented:
Ok I resolved conflicts

borisdayma commented:
Hi, just checking if i need to do anything else on my side?

borisdayma commented:
Hey guys, just checking on the progress here.
There's a few issues linked to this PR so I'd love to know what is still left to do.

borisdayma commented:
Thanks guys!!!
